Source/Core/DolphinQt2/Config/Graphics/PostProcessingConfigWindow.cpp

```diff
@@ -0,0 +1,359 @@
+
+#include <vector>
+
+#include <QCheckBox>
+#include <QDialog>
+#include <QDialogButtonBox>
+#include <QGridLayout>
+#include <QLabel>
+#include <QSlider>
+#include <QTabWidget>
+#include <QVBoxLayout>
+#include <QWidget>
+
+#include "VideoCommon/PostProcessing.h"
+#include "VideoCommon/RenderBase.h"
+#include "VideoCommon/VideoConfig.h"
+
+using ConfigurationOption = PostProcessingShaderConfiguration::ConfigurationOption;
+using OptionType = ConfigurationOption::OptionType;
+
+PostProcessingConfigWindow::PostProcessingConfigWindow(EnhancementsWidget* parent,
+                                                       const std::string& shader)
+    : QDialog((QWidget*)parent), m_shader(shader)
+{
+  if (g_renderer && g_renderer->GetPostProcessor())
+  {
+    m_post_processor = g_renderer->GetPostProcessor()->GetConfig();
+  }
+  else
+  {
+    m_post_processor = new PostProcessingShaderConfiguration();
+    m_post_processor->LoadShader(m_shader);
+  }
+
+  setWindowTitle(tr("Post Processing Shader Configuration"));
+  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+
+  PopulateGroups();
+  Create();
+}
+
+PostProcessingConfigWindow::~PostProcessingConfigWindow()
+{
+  m_post_processor->SaveOptionsConfiguration();
+  if (!(g_renderer && g_renderer->GetPostProcessor()))
+  {
+    delete m_post_processor;
+  }
+}
+
+void PostProcessingConfigWindow::PopulateGroups()
+{
+  const PostProcessingShaderConfiguration::ConfigMap& config_map = m_post_processor->GetOptions();
+
+  auto config_groups = std::vector<std::unique_ptr<ConfigGroup>>();
+  for (const auto& it : config_map)
+  {
+    std::unique_ptr<ConfigGroup> config_group = std::make_unique<ConfigGroup>(&it.second);
+    m_config_map[it.first] = config_group.get();
+    config_groups.push_back(std::move(config_group));
+  }
+
+  for (auto& config_group : config_groups)
+  {
+    const std::string& parent_name = config_group->GetParent();
+    if (parent_name.empty())
+    {
+      m_config_groups.emplace_back(std::move(config_group));
+    }
+    else
+    {
+      m_config_map[parent_name]->AddSubGroup(std::move(config_group));
+    }
+  }
+}
+
+void PostProcessingConfigWindow::Create()
+{
+  m_tabs = new QTabWidget();
+  auto* general = new QWidget(m_tabs);
+  auto* general_layout = new QGridLayout(general);
+
+  unsigned int row = 0;
+  bool add_general_page = false;
+  for (const auto& it : m_config_groups)
+  {
+    if (it->HasSubGroups())
+    {
+      auto* tab = CreateDependentTab(it);
+      m_tabs->addTab(tab, QString::fromStdString(it->GetGUIName()));
+    }
+    else
+    {
+      if (!add_general_page)
+      {
+        add_general_page = true;
+      }
+      row = it->AddWidgets(this, general_layout, row);
+    }
+  }
+
+  if (add_general_page)
+  {
+    m_tabs->insertTab(0, general, tr("General"));
+  }
+
+  m_buttons = new QDialogButtonBox(QDialogButtonBox::Ok);
+  connect(m_buttons, &QDialogButtonBox::accepted, this, &PostProcessingConfigWindow::accept);
```

Comment:
Please move this into a separate function called ``ConnectWidgets`` (or something similar). 
Having this separate from the UI initialization helps readability.